knowledge-base/changes-on-items-in-report-events-are-not-applied.md

```diff
@@ -0,0 +1,97 @@
+{
+    this.DataSource = GetReportData();
+    this.table.DataSource = GetTableData();
+}
+```
+  
+**As of [R3 2016](../upgrade-path-2016-r3#api-breaking-changes)**, we applied an optimization change in the report rendering algorithm and now each definition item property values are read and cached once the report rendering starts.   
+
+If you are working with item definitions in events, they will **not** be modified. For more details, check the updated [Understanding Events](../understanding-events)help article.
+
+Modifying report in events **after R3 2016:**  
+
+```cs
+void report_NeedDataSource(object sender, EventArgs e)
+{
+    Telerik.Reporting.Processing.Report processingReport = (Telerik.Reporting.Processing.Report)sender;
+    processingReport.DataSource = GetReportData();
+}
+```
+
+  
+> **Note**
+> <br/>
+> In the report **NeedDataSource** and **ItemDataBinding** events, the processing instances of the other report items are not created yet. Hence, it is not possible to modify other items in these Report events.  
+ 
+Modifying other items in events **after R3 2016:**  
+
+```cs
+void table_NeedDataSource(object sender, EventArgs e)
+{
+    Telerik.Reporting.Processing.Table processingTable = (Telerik.Reporting.Processing.Table)sender;
+    processingTable.DataSource = GetTableData();
+}
+```
+
+## Solution
+
+In general, in all versions of Telerik Reporting, changes in events are possible if they are applied on the **processing elements**, not the items' definitions (report elements accessible through **this.** in C# or **Me.** in VB.NET). The valid approaches are illustrated in [Accessing Report Items Programmatically](../faq-accessing-items).  
+  
+The recommended approach is to use the *expression engine* and to avoid custom code in events:  
+
+1. [Conditional Formatting](../styling-conditional-formatting) to control styles via expressions;
+2. [Bindings](../expressions-bindings) to set properties of items via expressions;
+3. [User functions](../expressions-user-functions) to set properties of items, when the [built-in functions](../expressions-reference) are not sufficient.
+
+The usage of the expression engine assures items will be modified at the right moment.  
+  
+Interesting help articles related to binding data via expressions and modifying reports:  
+
+- [How to use the ReportItem.DataObject property in expressions](../data-items-how-to-use-data-object)
+- [How to Databind to Collection Properties](./how-to-databind-to-collection-properties)
+- [How to bind Sub Report to Main Report's Data](./how-to-bind-sub-report-to-main-report-s-data)
+- [Change Sub Report based on Main Report's Data](./change-sub-report-based-on-main-report-s-data)
+- [Changing the connection string dynamically according to runtime data](./changing-the-connection-string-dynamically-according-to-runtime-data) (the same approach can be used if you need to change data sources and to modify the report structure at run-time).
+
+> **Note:** 
```

Comment:
This can go in a Notes section. It will become more visible for the user and will also leave the solution section cleaner. https://wiki.progress.com/display/DTS/Knowledge+Base+Content+Standards#KnowledgeBaseContentStandards-How-ToArticles